src/node_io.h

```diff
@@ -0,0 +1,236 @@
+#ifndef SRC_NODE_IO_H_
+#define SRC_NODE_IO_H_
+
+#include "node.h"
+
+namespace node {
+
+namespace io {
+
+enum io_error {
+  // Source would overflow the provided buffer
+  IO_ERROR_OVERFLOW = -256,
+  // Source does not support FLAG_PEEK
+  IO_ERROR_NO_PEEK = -512,
+  // Attempt to pull from an already ended source
+  IO_ERROR_EOF = -768,
+  // Source will not copy data into caller provided buffer
+  IO_ERROR_MUST_NO_COPY = -1024,
+};
+
+enum io_pull_flags {
+  IO_PULL_FLAG_NONE = 0x0,
+  // Callback must be invoked synchronously (also sets the
+  // IO_PULL_FLAG_MUST_COPY and IO_PULL_FLAG_STRICT_LENGTH flags)
+  IO_PULL_FLAG_SYNC = 0xD,
+  // Peek at what data is available, do not actually read.
+  IO_PULL_FLAG_PEEK = 0x2,
+  // Puller considers the length given to be strict,
+  // Source must not overflow
+  IO_PULL_FLAG_STRICT_LENGTH = 0x4,
+  // Puller requires source to copy data into the provided buffer
+  // (also sets the IO_PULL_FLAG_STRICT_LENGTH flag)
+  IO_PULL_FLAG_MUST_COPY = 0xC
+};
+
+enum io_bind_flags {
+  IO_BIND_FLAG_NONE = 0x0,
+  // Signal the Sink immediately after bind
+  IO_BIND_FLAG_SIGNAL = 0x1,
+  // Source supports peek
+  IO_BIND_FLAG_PEEK = 0x2,
+  // Pull must be performed synchronously
+  IO_BIND_FLAG_SYNC = 0x3
+};
+
+enum io_unbind_flags {
+  IO_UNBIND_FLAG_NONE = 0x0,
+  IO_UNBIND_FLAG_IMMEDIATE = 0x1
+};
+
+enum io_pull_status {
+  IO_PULL_STATUS_OK = 0x0,
+  // After this call, wait before asking again
+  IO_PULL_STATUS_WAIT = 0x1,
```

Comment:
Isn't it better to just wait in the source and return when data is available?

If the sink needed to I suppose it could peek or unbind on a timeout, but it seems better to wait in the source for data if there will be some rather than polling?

Reply:
> Isn't it better to just wait in the source and return when data is available?

Not necessarily. In the http/2 implementation, for instance, the callback from nghttp2 is sync. If data is not currently available, the stream is put into a deferred state until nghttp2 is explicitly told that data is now available. That is precisely the kind of use case this is meant to support.

The way Pull is defined, we have a simple set of primitives that allow the following scenarios, which can be used based on the needs of the Source and Sink:

1. Sync Pull ... Source waits to return until data is available (Blocking-Pull)
2. Async Pull ... Source waits to call callback until data is available (Nonblocking-Pull)
3. Sync Pull With Deferred Polling ... Source returns immediately saying data is not yet available, tells caller to check back later
4. Sync Pull With Deferred Signal ... Source returns immediately saying data is not yet available, will let caller know later when data is available (see Sink::Signal)
4. Async Pull With Deferred Polling (like above but using callback)
5. Async Pull With Deferred Signal (like above but using callback)

Reply:
> the stream is put into a deferred state until nghttp2 is explicitly told that data is now available. That is precisely the kind of use case this is meant to support.

So what you're saying is "4. Sync Pull With Deferred Signal"? i.e. a notification push to the sink? That would make sense but I was under the assumption we were avoiding that flow more entirely. I'll try it in the JS impl when I can.